ownCloud/Client/Actions/Actions+Extensions/MarkupAction.swift

```diff
@@ -0,0 +1,112 @@
+		} else if forContext.items.count > 1 {
+			return .none
+		} else if let item = forContext.items.first {
+			if let mimeType = item.mimeType {
+				if supportedMimeTypes.filter({
+					return mimeType.contains($0)
+				}).count == 0 {
+					return .none
+				}
+			} else {
+				return .none
+			}
+		}
+		// Examine items in context
+		return .middle
+	}
+
+	// MARK: - Action implementation
+	override func run() {
+		guard context.items.count > 0, let hostViewController = context.viewController, let core = self.core else {
+			self.completed(with: NSError(ocError: .insufficientParameters))
+			return
+		}
+
+		let hudViewController = DownloadItemsHUDViewController(core: core, downloadItems: context.items) { [weak hostViewController] (error, files) in
+			if let error = error {
+				if (error as NSError).isOCError(withCode: .cancelled) {
+					return
+				}
+
+				let appName = OCAppIdentity.shared.appName ?? "ownCloud"
+				let alertController = ThemedAlertController(with: "Cannot connect to ".localized + appName, message: appName + " couldn't download file(s)".localized, okLabel: "OK".localized, action: nil)
+
+				hostViewController?.present(alertController, animated: true)
+			} else {
+				guard let files = files, files.count > 0, let viewController = hostViewController else { return }
+				if let fileURL = files.first?.url, let item = self.context.items.first {
+					// Make sure self is around until interactionControllerDispatchGroup.leave() is called by the documentInteractionControllerDidDismissOptionsMenu delegate method implementation
+					self.interactionControllerDispatchGroup = DispatchGroup()
+					self.interactionControllerDispatchGroup?.enter()
+
+					self.interactionControllerDispatchGroup?.notify(queue: .main, execute: {
+						self.interactionController?.delegate = nil
+						self.interactionController = nil
+					})
+
+					let editDocumentViewController = EditDocumentViewController(with: fileURL, item: item, core: self.core)
+					editDocumentViewController.editDelegte = self
+					let navigationController = ThemeNavigationController(rootViewController: editDocumentViewController)
+					navigationController.modalPresentationStyle = .overFullScreen
+					viewController.present(navigationController, animated: true)
+				}
+			}
+		}
+
+		hudViewController.presentHUDOn(viewController: hostViewController)
+
+		self.completed()
+	}
+
+	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
+		if location == .moreItem || location == .moreFolder {
+			return UIImage(systemName: "pencil.tip.crop.circle")?.tinted(with: Theme.shared.activeCollection.tintColor)
+		}
+
+		return nil
+	}
+}
+
+@available(iOS 13.0, *)
+extension MarkupAction : EditDocumentViewControllerDelegate {
+	func editDocumentViewControllerDidDismiss(_ controller: EditDocumentViewController) {
+		interactionControllerDispatchGroup?.leave() // We're done! Trigger notify block and then release last reference to self.
```

Comment:
@hosy you could just move lines 80-81 in here instead of using dispatch group notify, right?